saleor/graphql/invoice/mutations.py

```diff
@@ -0,0 +1,274 @@
+        model = models.Invoice
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = InvoiceError
+        error_type_field = "invoice_errors"
+
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        invoice = cls.get_instance(info, **data)
+        response = super().perform_mutation(_root, info, **data)
+        events.invoice_deleted_event(user=info.context.user, invoice_id=invoice.pk)
+        return response
+
+
+class UpdateInvoiceInput(graphene.InputObjectType):
+    number = graphene.String(description="Invoice number")
+    url = graphene.String(description="URL of an invoice to download.")
+
+
+class UpdateInvoice(ModelMutation):
+    class Arguments:
+        id = graphene.ID(required=True, description="ID of an invoice to update.")
+        input = UpdateInvoiceInput(
+            required=True, description="Fields to use when updating an invoice."
+        )
+
+    class Meta:
+        description = "Updates an invoice."
+        model = models.Invoice
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = InvoiceError
+        error_type_field = "invoice_errors"
+
+    @classmethod
+    def clean_input(cls, info, instance, data):
+        number = instance.number or data["input"].get("number")
+        url = instance.external_url or data["input"].get("url")
+        if not number or not url:
+            raise ValidationError(
+                {
+                    "invoice": ValidationError(
+                        "URL and number need to be set after update operation.",
+                        code=InvoiceErrorCode.URL_OR_NUMBER_NOT_SET,
+                    )
+                }
+            )
+        return data["input"]
+
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        instance = cls.get_instance(info, **data)
+        cleaned_input = cls.clean_input(info, instance, data)
+        instance.update_invoice(
+            number=cleaned_input.get("number"), url=cleaned_input.get("url")
+        )
+        instance.status = JobStatus.SUCCESS
+        instance.save(update_fields=["external_url", "number", "updated_at", "status"])
+        return UpdateInvoice(invoice=instance)
+
+
+class SendInvoiceEmail(ModelMutation):
+    class Arguments:
+        id = graphene.ID(required=True, description="ID of an invoice to be sent.")
+
+    class Meta:
+        description = "Send an invoice by email."
+        model = models.Invoice
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = InvoiceError
+        error_type_field = "invoice_errors"
+
+    @classmethod
+    def clean_instance(cls, info, instance):
+        error_code = None
+
+        if instance.status != JobStatus.SUCCESS:
+            error_code = InvoiceErrorCode.NOT_READY
+        elif not instance.external_url or not instance.number:
+            error_code = InvoiceErrorCode.URL_OR_NUMBER_NOT_SET
```

Comment:
User shall not get this error because this is triggered by dashboard which implementation will show only ready invoices with send possibility. Also I doubt if anyone while building their own frontend would like to send not-ready invoices, therefore if someone sees this error there is something wrong with his system and that requires debugging the root cause anyway.

Reply:
Don't you think that for debugging, two different errors would be better than one for two causes?

Reply:
It's exactly same because both fields are in the same place so you see at first glance what is not set. Anyway, all right, explicit is better than implicit.